### PR TITLE
Add business rules manager

### PR DESCRIPTION
**Pull request by itkg-nanne**
https://github.com/open-orchestra/open-orchestra-cms-bundle/pull/2187
https://github.com/open-orchestra/open-orchestra-media-admin-bundle/pull/370
https://github.com/open-orchestra/open-orchestra/pull/1174

### CODE REVIEW

File: BackofficeBundle/Controller/ContentController.php [hunk 1]
```diff
@@ -33,15 +33,15 @@ class ContentController extends AbstractAdminController
...
-        $siteId = $this->get('open_orchestra_backoffice.context_manager')->getCurrentSiteId();
-        $site = $this->get('open_orchestra_model.repository.site')->findOneBySiteId($siteId);
-
         $content = $this->get('open_orchestra_model.repository.content')->findOneByLanguageAndVersion($contentId, $language, $version);
         if (!$content instanceof ContentInterface) {
             throw new \UnexpectedValueException();
         }
         $contentType = $this->get('open_orchestra_model.repository.content_type')->findOneByContentTypeIdInLastVersion($content->getContentType());
-        if (!$contentType instanceof ContentTypeInterface || !in_array($contentType->getContentTypeId(), $site->getContentTypes())) {
+        if (!$contentType instanceof ContentTypeInterface) {
```
Comment (alavieille) [inline BackofficeBundle/Controller/ContentController.php]: you can factorize this two if

**Comment by itkg-nanne** [comment]
update done @alavieille

File: Backoffice/BusinessRules/Strategies/BlockStrategy.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * @param BlockInterface $block
+     * @param array          $parameters
+     *
+     * @return boolean
+     */
+    public function canDelete(BlockInterface $block, array $parameters)
+    {
+        return 0 == $this->nodeRepository->countBlockUsed($block->getId()) && (array_key_exists('isTransverse', $parameters) ? $parameters['isTransverse'] : true);
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/BlockStrategy.php]: you can optimize this method and count only when block is transverse

File: Backoffice/BusinessRules/Strategies/NodeStrategy.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * @param NodeInterface $node
+     * @param array         $parameters
+     *
+     * @return boolean
+     */
+    public function canDelete(NodeInterface $node, array $parameters)
+    {
+        $siteId = $this->contextManager->getCurrentSiteId();
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/NodeStrategy.php]: why not use siteId in node ?

File: Backoffice/BusinessRules/Strategies/NodeStrategy.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * @param NodeInterface $node
+     * @param array            $parameters
+     *
+     * @return boolean
+     */
+    public function canDeleteVersion(NodeInterface $node, array $parameters)
+    {
+        return !$node->getStatus()->isPublishedState();
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/NodeStrategy.php]: and node should have one version

File: Backoffice/BusinessRules/Strategies/ContentStrategy.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * @param ContentInterface $content
+     * @param array            $parameters
+     *
+     * @return boolean
+     */
+    public function canDeleteVersion(ContentInterface $content, array $parameters)
+    {
+        return $this->isContentOnSiteAllowed($content) && !$content->getStatus()->isPublishedState();
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/ContentStrategy.php]: and content should have one version

**Comment by itkg-nanne** [comment]
update done @alavieille

File: ApiBundle/Controller/NodeController.php [hunk 6]
```diff
@@ -612,22 +613,18 @@ public function deleteNodeVersionsAction(Request $request, $nodeId, $language)
...
+                $this->get('open_orchestra_backoffice.business_rules_manager')->isGranted(NodeStrategy::DELETE_VERSION, $node)
+            ) {
+                $this->get('open_orchestra_backoffice.manager.node')->deleteBlockInNode($node);
+                $nodeIds[] = $node->getId();
+                $this->dispatchEvent(NodeEvents::NODE_DELETE_VERSION, new NodeEvent($node));
             }
-            $nodeRepository->removeNodeVersions($nodeIds);
-            $this->get('object_manager')->flush();
         }
+        $nodeRepository->removeNodeVersions($nodeIds);
```
Comment (alavieille) [inline ApiBundle/Controller/NodeController.php]: check if nodeIds isn't empty before call removeNodeVersions and flush

File: ApiBundle/Controller/ContentController.php [hunk 7]
```diff
@@ -178,20 +177,16 @@ public function deleteContentVersionsAction(Request $request, $contentId, $langu
...
+        foreach ($contents as $content) {
+            if ($this->isGranted(ContributionActionInterface::DELETE, $content) &&
+                $this->get('open_orchestra_backoffice.business_rules_manager')->isGranted(ContentStrategy::DELETE_VERSION, $content)
+            ) {
+                $storageIds[] = $content->getId();
+                $this->dispatchEvent(ContentEvents::CONTENT_DELETE_VERSION, new ContentEvent($content));
             }
-            $this->get('open_orchestra_model.repository.content')->removeContentVersion($storageIds);
         }
+        $this->get('open_orchestra_model.repository.content')->removeContentVersion($storageIds);
```
Comment (alavieille) [inline ApiBundle/Controller/ContentController.php]: check if storageIds isn't empty before

File: Backoffice/BusinessRules/Strategies/BlockStrategy.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+use OpenOrchestra\ModelInterface\Model\BlockInterface;
+use OpenOrchestra\Backoffice\Security\ContributionActionInterface;
+use OpenOrchestra\ModelInterface\Repository\NodeRepositoryInterface;
+
+/**
+ * class BlockStrategy
+ */
+class BlockStrategy extends AbstractBusinessRulesStrategy
+{
+
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/BlockStrategy.php]: remove this line

File: Backoffice/BusinessRules/Strategies/NodeStrategy.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace OpenOrchestra\Backoffice\BusinessRules\Strategies;
+
+use  OpenOrchestra\ModelInterface\Model\NodeInterface;
+use OpenOrchestra\Backoffice\Context\ContextManager;
+use OpenOrchestra\Backoffice\Security\ContributionActionInterface;
+use OpenOrchestra\ModelInterface\Repository\NodeRepositoryInterface;
+use OpenOrchestra\ModelInterface\Repository\SiteRepositoryInterface;
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/NodeStrategy.php]: remove this use

File: Backoffice/BusinessRules/Strategies/NodeStrategy.php [hunk 10]
```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace OpenOrchestra\Backoffice\BusinessRules\Strategies;
+
+use  OpenOrchestra\ModelInterface\Model\NodeInterface;
+use OpenOrchestra\Backoffice\Context\ContextManager;
```
Comment (alavieille) [inline Backoffice/BusinessRules/Strategies/NodeStrategy.php]: remove this use

**Pull request merged**